File: leetcode/py/1_two_sum.py
"""
Given an array of integers, return indices of the two numbers such that they add up to a specific target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

Example:

Given nums = [2, 7, 11, 15], target = 9,

Because nums[0] + nums[1] = 2 + 7 = 9,
return [0, 1].
"""

def solution_0(nums, target):
    for i in range(len(nums)):
        for j in range(i + 1, len(nums)):
            n1 = nums[i]
            n2 = nums[j]
            if n1 + n2 == target:
                return [i, j]

def solution_1(nums, target):
    lookup = {}
    for i, n in enumerate(nums):
        if target - n in lookup:
            return [i,lookup[target-n]]
        else:
            lookup[n] = i

print(solution_0([2,7,11,15], 9))
print(solution_1([2,7,11,15], 9))

File: leetcode/1_two_sum/1.two-sum.js
/*
 * @lc app=leetcode id=1 lang=javascript
 *
 * [1] Two Sum
 *
 * https://leetcode.com/problems/two-sum/description/
 *
 * algorithms
 * Easy (44.67%)
 * Total Accepted:    2.2M
 * Total Submissions: 5M
 * Testcase Example:  '[2,7,11,15]\n9'
 *
 * Given an array of integers, return indices of the two numbers such that they
 * add up to a specific target.
 * 
 * You may assume that each input would have exactly one solution, and you may
 * not use the same element twice.
 * 
 * Example:
 * 
 * 
 * Given nums = [2, 7, 11, 15], target = 9,
 * 
 * Because nums[0] + nums[1] = 2 + 7 = 9,
 * return [0, 1].
 * 
 */
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
	let map = new Map();
	let ret = [];
	for (let i = 0; i < nums.length; i++) {
		let v1 = nums[i];
		let v2 = target - v1;
		if (!map.has(v2)) {
			map.set(v1, i);
		} else {
			ret.push(map.get(v2));
			ret.push(i);
		}
	}
	return ret;
};
